RavenDB-20653 Reducing Corax entries usage

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20653

https://issues.hibernatingrhinos.com/issue/RavenDB-20228
https://issues.hibernatingrhinos.com/issue/RavenDB-20504
https://issues.hibernatingrhinos.com/issue/RavenDB-20435
https://issues.hibernatingrhinos.com/issue/RavenDB-20849
https://issues.hibernatingrhinos.com/issue/RavenDB-20862
https://issues.hibernatingrhinos.com/issue/RavenDB-20861
https://issues.hibernatingrhinos.com/issue/RavenDB-20761
https://issues.hibernatingrhinos.com/issue/RavenDB-20296
https://issues.hibernatingrhinos.com/issue/RavenDB-20635

### Additional description

Moving away from Corax entries to just the entries terms
That means that we can use less space and still get similar functionality.

So far migrated deletes, more like this and unary operations. 

Starting PR now to see how slow tests would behave.

ayende committed:
RavenDB-20504: Fixed the case where we are storing more than 1024 items.

ayende committed:
RavenDB-20653 Create a new format for indexed entry, instead of keeping track of the entire entry, we'll hold just the term ids for the entry
Those are encoded using varint, so should be very small. Estimating that we should drop from 100s of bytes per entry to < 16 in the common case.

ayende committed:
RavenDB-20653 Store the owner tree in the metadata for the terms, so we can go from the term to the owning field

ayende committed:
RavenDB-20653
- Build the index entry with term container id and the long / double values as well
- Properly handle deletions of entries by reading the new entry format

ayende committed:
RavenDB-20653
- Fixing allocated memory size in NativeList
- Generate the same id on Update again

redknightlois committed:
RavenDB-20635: Cleanup of what it is left of the back and forth for compact key handling

ayende committed:
RavenDB-20653
- Handle numeric fields lookup
- Restore handling of Suggestions
- Disabled update optimizations

ayende committed:
RavenDB-20653
- Refactored entries terms formatting and reading

ayende committed:
RavenDB-20653 Started refactoring away from Corax entries to entries terms

ayende committed:
RavenDB-20653 Migrating MoreLikeThis away from entries to entries terms

ayende committed:
RavenDB-20653
- Now recording the term frequencies in the entries terms
- Refactoring how we are handling MoreLikeThis

ayende committed:
RavenDB-20653 Major refactoring on how we are actually indexing in Corax
Now, calling Index() will just copy the data to be indexed and register that in the writer, instead of indexing immediately.
This is required so we'll be able to index the same document multiple times in the same batch, which is a feature we require in some rare cases.
Changed the actual behavior to Prepare & Commit, each of which is dong its own separate action

redknightlois committed:
RavenDB-20435: Allow to train bigger dictionaries because we will have just one per index.

arekpalinski commented:
Can you please check: https://jenkins.hibernatingrhinos.com/job/PR_Tests/6948/testReport/junit/SlowTests.Issues/RavenDB_19016/Can_Index_Nested_Document_Change_Same_Collection_options___DatabaseMode___Single___SearchEngineMode___Corax_/

arekpalinski commented on src/Corax/Queries/MultiUnaryMatch.cs:
```diff
@@ -387,141 +389,46 @@ public int Fill(Span<long> matches)
-                matches[currentIdx++] = matches[i];
-                // Its a match
-            }
+            Match:
```
Warning here

ayende committed:
RavenDB-20653 Test can now support Corax as well

ayende commented:
Yes, that was a pretty nasty one.
It passed previously by utter accident

Had to make a major change in how we are handling indexing in Corax, because we may actually get to the point where we are indexing the *same* document multiple times in the same indexing batch

ayende committed:
RavenDB-20653
Refactoring of how we handle reading of entries terms, including the ability to hold spatial data and stored fields
Major refactoring of how we are handling Spatial queries. We now register them in the entries terms, and can look them up cheaply

redknightlois committed:
RavenDB-20504: Applied review request.

redknightlois committed:
RavenDB-20504: Naive implementation of getting map objects to create dictionary

redknightlois committed:
RavenDB-20504: Improvements to the implementation readability

redknightlois committed:
RavenDB-20504: Handle configuration for how many objects to map in order to build dictionary

maciejaszyk commented on src/Corax/IndexWriter.cs:
```diff
@@ -435,71 +449,93 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
+            {
+                container.SingleItem = entryId;
+            }
+            else
+            {
+                container.Items = new List<long>
+                {
+                    container.SingleItem, 
+                    entryId,
+                };
+            }
```
It's possible to hit that block twice? It can override previous values then, right?

ayende replied:
You are correct, that is a problem then.

maciejaszyk commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -447,13 +460,13 @@ public bool Match(LazyStringValue other)
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public Span<byte> ToSpan()
+        public readonly Span<byte> ToSpan()
```
Warning:
```
Call to non-readonly member 'ByteString.Ptr.get' from a 'readonly' member results in an implicit copy of 'this'.
```

arekpalinski commented:
@ayende 
1. Can you check two warnings we have here? 
2. There is compilation issue.
3. Can you rebase and solve the conflict?

ayende committed:
RavenDB-20653 Refactoring how we configure fields in Corax to allow to define Storage of fields

ayende committed:
RavenDB-20653 Getting ready to handle stored fields in entries terms

ayende committed:
RavenDB-20653 Moving sharding to use Corax's new entries terms

ayende committed:
RavenDB-20653 Moving Corax Facets to use entries terms

ayende committed:
RavenDB-20653 Changing how we handle projections from java script to using Corax stored fields

ayende committed:
RavenDB-20653 Stored fields should have a dedicated container

ayende committed:
RavenDB-20653 Ensure that we only read the terms for stored fields when we *mean* to

ayende committed:
RavenDB-20653 Supporting stored fields that are arrays or numbers, supporting dynamic stored fields

ayende committed:
RavenDB-20653 Updating index entries to use entries terms

ayende committed:
RavenDB-20653
- Moving FieldsCache to be an explicitly term
- Projections of stored fields from Corax indexes now use the stored fields (updated tests that expected store by default)

ayende committed:
RavenDB-20653 Marking the key() field in map/reduce as stored, changing how we load the map/reduce object to use stored field

ayende committed:
RavenDB-20653 removing GetEntryReaderFor from the fast tests

ayende committed:
RavenDB-20653 removing GetEntryReaderFor() from slow tests and stress tests

ayende committed:
RavenDB-20653
- Renaming field to be clearer what it is for
- Better comments explaining what is going on
- Handle ordering of multiple values per term in sharding

ayende committed:
RavenDB-20653 Finalizing rebase

ayende committed:
RavenDB-20653 Fixing removal of terms that has frequency > 1

ayende commented on src/Corax/IndexWriter.cs:
```diff
@@ -1171,228 +1127,167 @@ void ExactInsert(ReadOnlySpan<byte> value)
-
-                for (int i = 0; i < tokenSpace.Length; i++)
+                //TODO: Maciej - we need to modify EntriesModifications to accept it externally 
+                for (int i = 0; i < reader.Frequency; i++)
```
@maciejaszyk Note this one

ayende committed:
RavenDB-20653 We'll no longer store entries at all in Corax!

ayende committed:
RavenDB-20653 Fixing warnings

ayende committed:
RavenDB-20653 Optimize look ups for deletion of old records

ayende commented on src/Corax/Queries/UnaryMatch.Ordinal.cs:
```diff
@@ -418,49 +298,23 @@ private static int FillFuncAllNonNull(ref UnaryMatch<TInner, TValueType> match,
-                        if (type.HasFlag(IndexEntryFieldType.HasNulls) && !type.HasFlag(IndexEntryFieldType.Empty))
+                        if(reader.FieldRootPage != fieldRoot)
+                            continue;
+                        if (reader.TermId == -1) // TODO: this is wrong, need to figure out what this looks like 
```
How do we do a *not* match? There are no tests for this?

arekpalinski replied:
@redknightlois or @maciejaszyk  ?

ayende commented on src/Raven.Server/Config/Categories/IndexingConfiguration.cs:
```diff
@@ -404,6 +404,12 @@ public virtual bool RunInMemory
+        [DefaultValue(100000)]
+        [IndexUpdateType(IndexUpdateType.Reset)]
+        [ConfigurationEntry("Indexing.Corax.DocumentsLimitForCompressionDictionaryCreation", ConfigurationEntryScope.ServerWideOrPerDatabaseOrPerIndex)]
+        public int DocumentsLimitForCompressionDictionaryCreation { get; protected set; }
```
https://github.com/ravendb/ravendb/pull/16687#discussion_r1246230317

ayende committed:
RavenDB-20635 Finishing rebase of dictionary branch

ayende committed:
RavenDB-20635 Removing unused generic argument

ayende committed:
RavenDB-20635 Avoid going to the dictionary twice

ayende committed:
RavenDB-20635 Encode the key once one lookup

redknightlois committed:
RavenDB-20435: Adjust the size of the buffer in cases the source is bigger.

This fixes the AVE caused while creating the index.

redknightlois committed:
RavenDB-20635: Optimization for the lookup of the interval frequency.

maciejaszyk commented on src/Corax/IndexWriter.cs:
```diff
@@ -435,81 +452,107 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
-            
             var entryId = ++_lastEntryId;
-            _entryIdToOffset.Add(entryId, entryContainerId);
+            _entriesAllocator.Allocate(key.Length, out var keyStr);
+            key.CopyTo(keyStr.ToSpan());
+            _entriesAllocator.Allocate(data.Length, out var dataStr);
+            data.CopyTo(dataStr.ToSpan());
+            _bufferedIndexedEntries[entryId] = dataStr;
```
Can we share that allocator with DocumentConverter/Writer? Since the converter is also using native memory we could save some time on copying data but I'm not sure if it's worth it.

ayende replied:
Ideally, yes. I'm going to be updating a lot of this code shortly, so not worth it at this point

ayende committed:
RavenDB-20635 Avoid doing another tree lookup for adding a term since we already just did one to find if we know what the term is

redknightlois committed:
RavenDB-20635: Use assembly baked literals.

maciejaszyk commented:
Please check the tests.

redknightlois committed:
RavenDB-20435: Using index output to train the dictionaries.

ayende committed:
RavenDB-20435 Minor refactoring

ayende committed:
RavenDB-20435 Started refactoring the way we are handling indexing to avoid creating index entries and then parsing them

ayende committed:
RavenDB-20435 Moving indexing functionality into the entry builder

ayende committed:
RavenDB-20435
- Fixing boost by adding that to the entry
- Fixing numeric terms indexing
- Fixing storing of blittables

ayende committed:
RavenDB-20435 Removing all references to Index(span, span) in favor of the entry builder

ayende committed:
RavenDB-20435 Removing IndexEntryWriter

ayende committed:
RavenDB-20435 Fixing handling of dynamic fields, now we update the query operation as soon as a new field is encountered

ayende committed:
RavenDB-20435 Fixing update of entries

ayende committed:
RavenDB-20435 Fixing tests

ayende committed:
RavenDB-20435 Changing the approach we use to handle duplicate entry indexing in the same batch, we'll flush the batch and start a *new* batch in that case

ayende committed:
RavenDB-20435 Removing more of the index entries format

ayende committed:
RavenDB-20653 Removing IndexEntry from the codebase

ayende committed:
RavenDB-20653 Reuse the CompactKey instance

ayende committed:
RavenDB-20653 Reuse the FastPForDecoder allocation to improve performance

ayende committed:
RavenDB-20653 Removing WriteScope and re-adding dictionary training

ayende committed:
RavenDB-20653 Handling analyzers for the dictionary training properly, handling nested blittable storage

ayende committed:
RavenDB-20653 Fixing reference usage

ayende committed:
RavenDB-20653 Don't double count the entries terms in the compact tree as well

redknightlois commented on src/Corax/IndexWriter.cs:
```diff
@@ -435,409 +454,462 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
-                var newFieldReader = newEntryReader.GetFieldReaderFor(fieldBinding.FieldId);
-                var oldFieldReader = oldEntryReader.GetFieldReaderFor(fieldBinding.FieldId);
-                if (oldType != newType)
+            void RegisterTerm(Slice fieldName, ReadOnlySpan<byte> term, StoredFieldType type)
```
Context is needed on how the Register Term -> Process term process works.

redknightlois commented on src/Corax/IndexWriter.cs:
```diff
@@ -490,15 +490,33 @@ public IndexWriter([NotNull] Transaction tx, IndexFieldsMapping fieldsMapping) :
 
-        public class IndexEntryBuilder : IDisposable
+        public interface IIndexEntryBuilder
+        {
```
What's the rationale behind the introduction of this interface?

ayende replied:
We need it for the training, we have two builders. One for training and one for indexing

redknightlois commented on test/FastTests/Corax/Bugs/FacetIndexingRepro.cs:
```diff
@@ -251,17 +251,13 @@ public unsafe void CanSuccessfullyIndexData()
-                        var fieldReader = reader.GetFieldReaderFor(i);
-                        fieldReader.Read(out Span<byte> s);
-                        indexEntryBuilder.Write(i, s);
+                        var len = br.Read7BitEncodedInt();
```
What are we writing and why?

ayende replied:
I ported a failing test that embedded the raw index entries, to remove that, I changed the format of the file and now we are reading it this way

redknightlois commented on test/FastTests/Corax/Bugs/FacetIndexingRepro.cs:
```diff
@@ -147,22 +147,19 @@ public unsafe void ShouldNotCorrupt()
-                        var fieldReader = reader.GetFieldReaderFor(i);
-                        fieldReader.Read(out Span<byte> s);
-                        indexEntryBuilder.Write(i, s);
+                       int len = br.Read7BitEncodedInt();
```
What are we writing and why.

ayende replied:
I ported a failing test that embedded the raw index entries, to remove that, I changed the format of the file and now we are reading it this way

redknightlois committed:
RavenDB-20635: Extend the Memory Compare routines.

redknightlois committed:
RavenDB-20635: Fully managed CompactKey

ayende committed:
RavenDB-20296 Added a way to account for _every single page_ in the system.
This is done so we can check why the storage report is out of whack, not meant for general consumption

ayende committed:
RavenDB-20296
- Take into account that the large posting list has inner values
- Remove the old dictionary when replacing it

ayende committed:
RavenDB-20296 We have to keep the field around, since the default dictionary already assumes it

ayende committed:
RavenDB-20296 Fixing Container size computation when there is a lot of delete traffic as well

redknightlois committed:
RavenDB-20635: Update the memory compare microbenchmark

ayende committed:
RavenDB-20761: Corax - restore update optimization

redknightlois committed:
RavenDB-20635: Enabled CompactKey sharing disabled by RavenDB-20143

ayende committed:
RavenDB-20761 We cannot use the _entriesAllocator here, since we reset is if we flush the batch

ayende committed:
RavenDB-20761 Restore update optimization for Corax

ayende committed:
RavenDB-20653 We no longer need PrepareAndcommit, using just Commit

ayende committed:
RavenDB-20861 Allow Corax.EntryModifications to accept a value and frequency in one shot

ppekrol commented:
Could you rebase this? There are some conflicts and current code does not compile as I can see.

ayende committed:
RavenDB-20862 Test what happens when Corax Update a document and the frequency of an item changes

ayende commented on src/Voron/Data/CompactTrees/CompactKey.cs:
```diff
@@ -17,12 +18,15 @@ namespace Voron.Data.CompactTrees;
 public unsafe class CompactKey : IDisposable
 {
-    public readonly LowLevelTransaction Owner;
+    private static readonly ArrayPool<byte> StoragePool = ArrayPool<byte>.Shared;
+    private static readonly ArrayPool<long> KeyMappingPool = ArrayPool<long>.Shared;
```
why do this? They are the same instances, after all

redknightlois replied:
One is long and the other one is bytes... I want to avoid having to Unsafe access the long array as size of byte.

ayende replied:
I was actually talking about why you are making a `static field` for each of those, where the _is_ ` static field` for them already in the `Shared`

ayende commented on src/Voron/Impl/LowLevelTransaction.cs:
```diff
@@ -1046,13 +1047,22 @@ public void FreePage(long pageNumber)
+            return key;
+        }
+
+        public void ReleaseCompactKey(ref CompactKey key)
```
shouldn't you handle this already being null?

redknightlois replied:
If that happens, it should fail fast. It's a big bug.

ayende replied:
In general, we aren't going to consider disposing twice a sign of a problem.
That can happen for a variety of reasons
For example:

![image](https://github.com/ravendb/ravendb/assets/116915/a19b86f8-5d4e-46bc-9684-212e77523376)

ayende replied:
In this case, we have a batch, and we hit a scenario where we have to flush it, then dispose.
The flush already dispose the key, but the dispose does that as well

One can also thing of other scenarios, where we have a `CompactKeyCacheScope` that wasn't init (maybe an error) but is called from the dispose.

ayende commented:
Yes, I'll handle this.
Currently making sure that this is all working properly with mine latest changes + federicos

ayende commented:
Rebased

ayende commented:
Looking at the test failures, nothing to do with Corax
Some studio things failed (missing `eslint`)

ayende committed:
RavenDB-20653 Better name for stored fields

ayende committed:
RavenDB-20228 Corax: Failing test RavenDB_19016.Can_Index_Nested_Document_Change_Different_Collections

ayende committed:
RavenDB-20268 Provide a way to detect that a field has multiple terms for a field (and thus cannot be used as streaming optimization)

ayende committed:
RavenDB-20268 Fixing the build on Linux

arekpalinski commented:
Latest test failures: https://jenkins.hibernatingrhinos.com/job/PR_Tests/7261/testReport/

ayende committed:
RavenDB-20268
- Match Lucene behavior when loading field value from index, do not require a stored field
- Adding better debug capabilities for entries terms

ayende committed:
RavenDB-20268 Removing bad attribute, cleaning namespaces

ayende committed:
RavenDB-20268 Fixing compilation after rebase

ayende committed:
RavenDB-20268 Fixing NRE

ayende committed:
RavenDB-20268
- Better handling for storing of JSON fields
- Better handling on dynamic fields that holds JSON data

ayende committed:
RavenDB-20268 Fixing projection of a single stored field that was stored as a list, will now also be projected as list

maciejaszyk commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexReadOperation.cs:
```diff
@@ -1157,7 +1162,7 @@ public override SortedSet<string> Terms(string field, string fromValue, long pag
                 {
                     builderParameters = new(IndexSearcher, _allocator, serverContext, context, query, _index, query.QueryParameters, QueryBuilderFactories,
-                        _fieldMappings, null, null /* allow highlighting? */, CoraxQueryBuilder.TakeAll, indexReadOperation: this, token: token);
+                        _fieldMappings, null, null /* allow highlighting? */, CoraxQueryBuilder.TakeAll, indexReadOperation: this);
```
Token should be passed into primitives, right?

ayende replied:
Yes

ayende committed:
RavenDB-20268 We shouldn't consider lists of dynamic fields as lists, so we'll ignore that

ayende committed:
RavenDB-20268 Resetting the reader so we won't miss values

ayende committed:
RavenDB-20268
- Allow to project empty lists from Corax
- Better entry terms debug information
- Explicitly record empty lists if they are stored
- Skip indexing if there have been no fields (except id()) that were indexed
- We no longer assume that we can always project from Corax
- Always project id() from corax indexes

ayende committed:
RavenDB-20268
- For dynamic fields, we reset the list behavior and start from scratch

ayende committed:
RavenDB-20268
- Ensuring that we use the right size when decoding a value in CompactKey
- Projections will now always use lower document id for consistency between Lucene and Corax
- Handle null/empty string projections
- Handle running indexing in test context
- Complex object that we weren't told _to_ index shouldn't cause an error

arekpalinski commented on src/Corax/IndexSearcher/IndexSearcher.cs:
```diff
@@ -54,35 +54,43 @@ public sealed unsafe partial class IndexSearcher : IDisposable
         _metadataTree = _transaction.ReadTree(Constants.IndexMetadataSlice);
-        _entryIdToOffset = _transaction.LookupFor<Int64LookupKey>(Constants.IndexWriter.EntryIdToOffsetSlice);
+        _entryIdToLocation = _transaction.LookupFor<Int64LookupKey>(Constants.IndexWriter.EntryIdToLocationSlice);
+        //TODO: Temporary workaround until we are done with single dic
```
What does this TODO mean exactly? What remains to be done? CC @redknightlois

ayende replied:
Already did that, need to remove the `TODO`

arekpalinski commented on src/Corax/IndexWriter.cs:
```diff
@@ -435,409 +396,496 @@ private IndexWriter(IndexFieldsMapping fieldsMapping)
-        /// <returns>Encoded id (with freq/container_type)</returns>
-        public unsafe long Index(Span<byte> data)
+
+        public class IndexEntryBuilder : IDisposable, IIndexEntryBuilder
```
We have a couple of classes / structs / interfaces defined inside `IndexWriter.cs` (EntriesModifications, TermInEntryModification, IndexedField, ...). Maybe it's time to move them to dedicated files? 

CC @maciejaszyk

ayende replied:
Would rather delay that after merging this

arekpalinski replied:
Yes, I agree. Better to handle that once we merge this.

arekpalinski replied:
@maciejaszyk can you look into that later?

arekpalinski commented on src/Corax/Queries/MultiUnaryMatch.cs:
```diff
@@ -387,141 +389,45 @@ public int Fill(Span<long> matches)
-            _ => comparer.CompareNumerical(iterator.Double)
+            MultiUnaryItem.DataType.Slice => comparer.CompareLiteral(iterator.Current.Decoded()),
+            MultiUnaryItem.DataType.Long => comparer.CompareNumerical(iterator.CurrentLong),
+            _ => comparer.CompareNumerical(iterator.CurrentDouble)
```
Why not handling `MultiUnaryItem.DataType.Double` explicitly but as the default option?

ayende replied:
done

arekpalinski commented on src/Corax/Queries/UnaryMatch.Ordinal.cs:
```diff
@@ -637,47 +452,24 @@ private static int FillFuncAnyNull(ref UnaryMatch<TInner, TValueType> match, Spa
-                        else if (type.HasFlag(IndexEntryFieldType.Tuple) || type.HasFlag(IndexEntryFieldType.Simple))
+                        if(reader.FieldRootPage != fieldRoot)
+                            continue;
+                        if (reader.TermId == -1) // TODO: this is wrong, need to fix it
```
Reminder about this TODO

ayende replied:
@maciejaszyk Can you write a test that would look at this scenario?
I couldn't find a way to trigger this.

arekpalinski commented on src/Corax/Utils/EntryTermsReader.cs:
```diff
@@ -0,0 +1,410 @@
+///
+/// * 0b010 - indicates a spatial point. The term id in this case holds the *field root page*, not a term. This is followed by two doubles (raw) representing lat/lng.
+/// * 0b110 - indicates that this is a  *stored* field. The term id contains the type of term we deal with, not a term.
+/// </summary>
```
Just to make it sure - this description is up to date according to the last commit in this PR, right?

ayende replied:
Yes, this is up to date

arekpalinski commented on src/Raven.Server/Documents/Queries/MoreLikeThis/Corax/MoreLikeThis.cs:
```diff
@@ -333,16 +199,33 @@ public override void SetMaxDocFreqPct(int maxPercentage)
+        void InsertTerm(ReadOnlySpan<byte> term, int freq)
+        {
+            var termAsString = Encoding.UTF8.GetString(term);
+            if (IsNoiseWord(termAsString)) // TODO optimize
```
Maybe let's create a ticket for that with more defails?

ayende replied:
The idea is that we should skip creating the string instance. Very low priority

arekpalinski commented on src/Voron/Util/NativeList.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Voron.Util;
+
+public unsafe struct NativeList<T> : IDisposable
```
Can we have some unit tests for that?

arekpalinski replied:
@maciejaszyk can you look into that later?

arekpalinski reviewed (commented):
LGTM. Although it's pretty hard to me to understand most of Corax code. Just reading it is pretty challenging. I didn't spot any mistakes or places that look fishy to me.

ayende committed:
RavenDB-20268
- Respect the "index empty entries" flag
- _all_stored_fields should now get only the stored fields
- Matching document id projection behavior when projecting from index between Lucene & Corax

ayende committed:
RavenDB-20268 Properly handle should skip for dynamic nulls

ppekrol commented on src/Corax/Mappings/IndexFieldsMappingBuilder.cs:
```diff
@@ -76,14 +72,14 @@ public IndexFieldsMappingBuilder AddDynamicBinding(Slice fieldName, FieldIndexin
     
     public IndexFieldsMappingBuilder AddBinding(int fieldId, string fieldName, Analyzer analyzer = null, bool hasSuggestion = false,
-        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false)
+        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool hasSpatial = false, bool shouldStore = false)
```
can we keep the same order of properties? shouldStore should be before hasSpatial

ayende replied:
What is the logic? What is the previous order of properties?

ppekrol replied:
In all the method in this class you have:
```
        FieldIndexingMode fieldIndexingMode = FieldIndexingMode.Normal, bool shouldStore = false, bool hasSpatial = false)
```

and in this one it s the other way around

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/AnonymousCoraxDocumentConverter.cs:
```diff
@@ -24,88 +22,68 @@ public abstract class AnonymousCoraxDocumentConverterBase : CoraxDocumentConvert
         {
-            ResetEntriesWriter();
-            throw;
+            var bjo = indexContext.ReadObject(storedValue, "corax field as json");
```
is there a place where we can dispose this?

ayende replied:
Now there is :-)

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs:
```diff
@@ -24,85 +23,71 @@ public class CoraxDocumentConverter : CoraxDocumentConverterBase
-                        fixed (byte* bPtr = blittableBuffer)
-                            document.Data.CopyTo(bPtr);
+                    fixed (byte* bPtr = blittableBuffer)
+                        document.Data.CopyTo(bPtr);
```
can we dispose document after this?

ayende replied:
That is coming from the caller, not for us to dispose, no?

ppekrol replied:
Ok

ayende committed:
RavenDB-20268 Matching Lucene behavior for inserting new documents with no fields

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentTrainEnumerator.cs:
```diff
@@ -0,0 +1,253 @@
+                    break;
+
+                var doc = itemEnumerator.Current.Item as Document;
+                if (doc == null)
```
when is this possible?

ayende replied:
Counters & Time Series,  which aren't relevent here for now

ppekrol replied:
Would be best to throw for now? or do a direct cast?

ayende replied:
done

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxIndexPersistence.cs:
```diff
@@ -111,10 +113,59 @@ public override void Clean(IndexCleanup mode)
+            var converter = CreateConverter(_index);
+            converter.IgnoreComplexObjectsDuringIndex = true; // for training, we don't care
+            var enumerator = new CoraxDocumentTrainEnumerator(indexContext, converter, _index, _index.Type, documentStorage, queryContext, _index.Collections, _index.Configuration.DocumentsLimitForCompressionDictionaryCreation);
+            var testEnumerator = new CoraxDocumentTrainEnumerator(indexContext, converter, _index, _index.Type, documentStorage, queryContext, _index.Collections, IndexingCompressionMaxTestDocuments);
```
IndexingCompressionMaxTestDocumen comes from a const, should this be configurable or I do not know e.g `_index.Configuration.DocumentsLimitForCompressionDictionaryCreation / 10` ?

ayende replied:
@redknightlois ?

ayende committed:
RavenDB-20268 PR comments

ppekrol commented on test/SlowTests/Issues/RavenDB-10343.cs:
```diff
@@ -13,9 +13,9 @@
 namespace SlowTests.Issues
 {
-    public class RavenDB_10343 : RavenTestBase
+    public class RavenDb10343 : RavenTestBase
```
:(

ayende replied:
reverted

ayende committed:
RavenDB-20268
- Do not attempt to project id() from a map/reduce
- Better handling of empty documents during indexing

ayende committed:
RavenDB-20268
- Adding better error message for a rare unreproducible exception

ayende committed:
RavenDB-20268 PR Comments

ayende committed:
RavenDB-20268 Better message for rare exception

ppekrol commented:
test this in debug please

redknightlois commented:
Damn, pushed to the wrong upstream. My bad, passes all fast tests, should I revert?

redknightlois commented:
This index on a small dataset:
```
from t in docs.Tweets
select new
{
    hashtag = t.entities.hashtags.Select(x=>x.text),
    user = t.user,
    t.created_at
}
```
Lucene: 2m14
Corax: 1m50

There are still a few adjustments that I have to do.

ppekrol commented:
Reverted back. Please check failing test here: https://jenkins.hibernatingrhinos.com/job/PR_Tests/7354/

ayende commented:
> This index on a small dataset:
> 
> ```
> from t in docs.Tweets
> select new
> {
>     hashtag = t.entities.hashtags.Select(x=>x.text),
>     user = t.user,
>     t.created_at
> }
> ```
> 
> Lucene: 2m14 Corax: 1m50
> 
> There are still a few adjustments that I have to do.

What were the initial numbers?

ayende commented:
Hurray!
